scripts/convert-ESO-PV-forecasts-from-CSV-to-HDF5.py

```diff
@@ -0,0 +1,195 @@
+    and the values are the pd.DataFrame of forecast data.
+    """
+    eso_forecasts_df.SCRIPT_NAME = eso_forecasts_df.SCRIPT_NAME.str.upper()
+
+    eso_forecasts = {}
+    for algo_name in ESO_ALGO_NAMES:
+        mask = eso_forecasts_df.SCRIPT_NAME.str.contains(f'_{algo_name}_')
+        eso_forecasts[algo_name] = eso_forecasts_df[mask]
+        eso_forecasts[algo_name] = eso_forecasts[algo_name].drop(columns='SCRIPT_NAME')
+
+    return eso_forecasts
+
+
+def convert_to_dataarray(df: pd.DataFrame) -> xr.DataArray:
+    """Convert forecast data to a 3-dimensional DataArray, with these dimensions:
+
+    1. `gsp_id`: A string identifying the Grid Supply Point region.
+    2. `forecast_date_time`: The UTC datetime when ESO ran their forecast.  Not exactly the
+        same as the NWP init time.  This script takes the floor('30T') of the
+        original forecast_date_time from ESO.
+    3. `step`: The Timedelta between the forecast_date_time and the target_date_time.
+    """
+    df = df.copy()  # So we don't modify the passed-in `df` object.
+
+    # The FORECAST_DATE_TIME is, I think, the time when ESO ran their forecasting script.
+    # The times are all over the place.  So floor the datetime to the nearest half-hour:
+    df['forecast_date_time'] = df['FORECAST_DATE_TIME'].dt.floor('30T')
+    df = df.drop(columns='FORECAST_DATE_TIME')
+
+    # Calculate the forecast "step":
+    df['step'] = df['TARGET_DATE_TIME'] - df['forecast_date_time']
+    df = df.drop(columns='TARGET_DATE_TIME')
+
+    # Make sure "step" is positive:
+    # (step can be negative for ASL forecasts, for some reasons).
+    df = df[df.step >= pd.Timedelta(0)]
+
+    # Rename to more column names more like the ones we're used to.
+    df = df.rename(columns={'SITE_ID': 'gsp_id'})
+
+    # Set index
+    df = df.set_index(['gsp_id', 'forecast_date_time', 'step'])
+    series = df.squeeze()
+    del df
+
+    # Remove duplicate indicies
+    duplicated = series.index.duplicated()
+    series = series[~duplicated]
+    return series.to_xarray()
+
+
+def main():
+    """Loop round all the files."""
+    filenames = filenames_and_datetime_periods(ESO_CSV_PATH)
+    n = len(filenames)
+    datasets = []
+
+    for i, filename in enumerate(filenames.values):
+        print(f'{i+1:2d}/{n:2d}: {filename}', flush=True)
+        eso_forecasts_df = load_csv(filename)
+        eso_forecasts_df = keep_only_2_day_forecasts(eso_forecasts_df)
+        split_forecasts = split_asl_and_ml_forecasts(eso_forecasts_df)
+        del eso_forecasts_df
+
+        data_arrays = {}
+        for algo_name in ESO_ALGO_NAMES:
+            data_arrays[algo_name] = convert_to_dataarray(split_forecasts[algo_name])
+        del split_forecasts
+
+        dataset = xr.Dataset(data_arrays)
+        datasets.append(dataset)
+        del dataset, data_arrays
+
+    dataset = xr.concat(datasets, dim='forecast_date_time')
+
+    dataset.to_netcdf(DST_NETCDF)
```

Comment:
can you say roughly how big this file is? If its very big, would any encoding or compression help?

Reply:
I've added a comment.